### PR TITLE
Offline download batches

### LINKED ISSUE #10252: Downloading an offline region on Android takes ~3x as long as iOS
**Platform:** iOS and Android
**Mapbox SDK version:** Android 5.2.0-beta.2, iOS 3.6

### Description of benchmark test

@boundsj and I attempted to download the same offline region using the Android and iOS versions of the SDK. We tested on both wifi (in the Mapbox SF office) and LTE, noting network speeds, download times, and final download sizes along the way. 

**iOS offline region**

```
    CLLocationCoordinate2D sw = CLLocationCoordinate2DMake(40.759096, -74.0024069);
    CLLocationCoordinate2D ne = CLLocationCoordinate2DMake(40.789221, -73.9398247);
    MGLCoordinateBounds bounds = MGLCoordinateBoundsMake(sw, ne);
...
    maximumZoomLevel = 22
    mapView.zoomLevel = 12.183443853857856
```

**Android offline region**

```
LatLngBounds latLngBounds = new LatLngBounds.Builder()
        .include(new LatLng(40.789221, -73.9398247)) // Northeast
        .include(new LatLng(40.759096, -74.0024069)) // Southwest
        .build();
...
OfflineTilePyramidRegionDefinition definition = new OfflineTilePyramidRegionDefinition(
        mapboxMap.getStyleUrl(),
        latLngBounds,
        12,
        22,
        MainActivity.this.getResources().getDisplayMetrics().density);
```

We both tested using the latest Mapbox Streets style (`mapbox://styles/mapbox/streets-v10`).

#### Mapbox WiFi
_Stats_
- Ping: 63ms
- Download: 118 Mbps
- Upload: 32.37 Mbps

🍏  **iOS**: 51 seconds for 78.9 MB
🤖  **Android**:  3 minutes, 57 seconds for 76.06 MB

#### LTE 
_Stats (iOS)_
- Network: AT&T
- Ping: 37ms
- Download: 26.36 Mbps
- Upload: 17.25 Mbps

_Stats (Android)_
- Network: T-Mobile
- Ping:  100 ms
- Download: 5.7 Mbps
- Upload:  15 Mbps

🍏  **iOS**: 53 seconds for 78.9 MB
🤖  **Android**: ~3 minutes, 20 seconds for 76.06 MB 

### Expected behavior

Download speeds are similar between platforms.

### Actual behavior

Download speeds are around 3 times longer on Android when compared to iOS.

Note: while the network speeds were significantly slower on our Android test device, the discrepancy in download times is consistent with those found while testing on the same WiFi network, suggesting that network speed is not the bottleneck.

### LINKED ISSUE #11217: Optimize the number of transactions for an offline download
Per internal conversation, operating fewer transactions for an offline download improves its speed. The tradeoff is that when a larger transaction fails, more offline data is lost. The purpose of this ticket is to find the optimal transaction number that balances speed and reliability. 

@kkaefer @ivovandongen , how can we strike the right balance here?

### PR DESCRIPTION
**Pull request by ivovandongen**
Fixes #11217
Closes #10252

Alternative to #11235 using batching of resource update/inserts.

Somewhat slower: +/- 30 seconds on Pixel XL 2, 130 seconds on Galaxy Nexus. Higher memory consumption; had to keep the batch size relatively small to avoid running out of memory on old devices.

### CODE REVIEW

File: platform/default/mbgl/storage/offline_database.cpp [hunk 1]
```diff
@@ -136,6 +136,12 @@ mapbox::sqlite::Statement& OfflineDatabase::getStatement(const char* sql) {
     return *it->second;
 }
 
+void OfflineDatabase::batch(BatchedFn&& fn) {
```
Comment (kkaefer) [inline platform/default/mbgl/storage/offline_database.cpp]: Instead of exposing a generic function like this, can we pass a list/collection of `Resource`/`Response` pairs and handle iteration over those, and transaction management internally? E.g. have a public `putRegionResources()` function that internally calls `putRegionResource` for every item, wrapping everything in a transaction.

**Comment by jfirebaugh** [comment]
I just discovered that you _can_ `REPLACE` in a way that avoids generating a new `id`:

```
REPLACE INTO resources (url, ..., id)
VALUES                 (?1,  ..., (SELECT id FROM resources WHERE url = ?1))
```

This takes advantage of the fact that `id` is the ROWID column, which will replace a NULL value with an autogenerated value on inserts.

However, there does not appear to be a way of determining whether such a statement replaced or inserted, which `putResource` / `putTile` needs for the return value. 😢If there were, we could avoid the transaction entirely.

**Comment by jfirebaugh** [comment]
@ivovandongen Can you benchmark this approach versus https://github.com/mapbox/mapbox-gl-native/compare/no-transaction? I'm not sure if small automatic deferred transactions, or batched immediate transactions will be faster.

**Comment by ivovandongen** [comment]
>@ivovandongen Can you benchmark this approach versus https://github.com/mapbox/mapbox-gl-native/compare/no-transaction? I'm not sure if small automatic deferred transactions, or batched immediate transactions will be faster.

@jfirebaugh Between 95 - 100 seconds on the Pixel XL 2. The Galaxy Nexus is still charging, but I don't think it will be much better I'm afraid.

**Comment by jfirebaugh** [comment]
I'm clear how 95-100 seconds compares. Can you summarize the results from the three approaches (`master`, `no-transation`, `offline-download-batches`)?

**Comment by ivovandongen** [comment]
|  | Pixel 2 | Galaxy Nexus |
--- |------------ | -------------
master |  ~90 seconds  | 15 minutes |
[no transaction](https://github.com/mapbox/mapbox-gl-native/compare/no-transaction) | 95 - 100 seconds | _not tested_ |
[offline-download-batches](https://github.com/mapbox/mapbox-gl-native/pull/11284) | ~30 seconds | ~130 seconds |
[longer running transactions](https://github.com/mapbox/mapbox-gl-native/pull/11235) | ~20 seconds | ~130 seconds |

Tested with 

- Berlin region: 5195 tiles, north:52.6780473464 east:13.7603759766 south:52.3305137868 west:13.0627441406 max-zoom:15 
- Warmed up cache on the edge location

**Comment by kkaefer** [comment]
\o/

**Comment by friedbunny** [comment]
iOS changelog added in https://github.com/mapbox/mapbox-gl-native/pull/12086.
